homeassistant/components/pi_hole/__init__.py

```diff
@@ -180,26 +108,88 @@ async def async_update_data():
         except HoleError as err:
             raise UpdateFailed(f"Failed to communicating with API: {err}")
 
+    entry.add_update_listener(async_update_listener)
+
     coordinator = DataUpdateCoordinator(
         hass,
         _LOGGER,
         name=name,
         update_method=async_update_data,
         update_interval=MIN_TIME_BETWEEN_UPDATES,
     )
-    hass.data[DOMAIN][name] = {
+    hass.data[DOMAIN][entry.entry_id] = {
         DATA_KEY_API: api,
         DATA_KEY_COORDINATOR: coordinator,
     }
 
-    hass.async_create_task(
-        hass.config_entries.async_forward_entry_setup(entry, SENSOR_DOMAIN)
-    )
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
 
     return True
 
 
+async def async_update_listener(hass, entry):
+    """Handle options update."""
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
 async def async_unload_entry(hass, entry):
-    """Unload pi-hole entry."""
-    hass.data[DOMAIN].pop(entry.data[CONF_NAME])
-    return await hass.config_entries.async_forward_entry_unload(entry, SENSOR_DOMAIN)
+    """Unload Pi-hole entry."""
+    hass.data[DOMAIN].pop(entry.entry_id)
+    return all(
+        [
+            await hass.config_entries.async_forward_entry_unload(entry, platform)
```

Comment:
Please use `asyncio.gather` to await the unload of the platforms concurrently.

Select the platforms based on api key set or not.